# Kvantni algoritmi 1

U ovom poglavlju opisaćemo neke od najvažnijih kvantnih algoritama i uporediti ih sa algoritmima koji rešavaju isti problem na klasičnom računaru. Da bi se što bolje iskoristile prednosti oba pristupa (klasičnog i kvantnog), zamišljeno je da se koriste i kvantni i klasični komunikacioni kanali, kao i kvantni i klasični algoritmi.

Kao meru složenosti algoritama, računaćemo broj kvantnih kola koja su potrebna da se izračuna rešenje, budući da broj kvantnih kola određuje broj transformacija koje je potrebno izvršiti. Kao i u klasičnom računaru, neke transformacije se smatraju osnovnim, a neke se izvode na osnovu drugih. Ipak, postoje razlike: na klasičnom računaru je moguće sve operacije izraziti preko $NAND$ operacija, dok na kvantnom računaru (za sad) nije pronađen takav konačan skup kvantnih kola.

Pokazano je da se operacijama iz skupa svih unitarnih transformacija nad jednim kubitom i transformacije $CNOT$ nad dva kubita mogu odrediti sve ostale unitarne transformacije. Ipak, nije realistično očekivati da se implementira svako kolo nad jednim kubitom, jer njih ima beskonačno mnogo \footnote{Na primer, za $t > 0$, matrice $\begin{pmatrix} 1 & 0 \\ 0 & e^{i \pi / t}\end{pmatrix}$ su unitarne i njih ima beskonačno mnogo.}. Dokazano je da se sve unitarne transformacije nad jednim ili dva kubita mogu aproksimirati onima iz skupa koji čine $CNOT$, Adamarova $H$ transformacija i promena faze $T_{\pi/8}$. Ako je greška aproksimacije $\epsilon$, broj kvantnih kola u aproksimaciji je polinom po $\log\epsilon$.

U nastavku ćemo podrazumevati da je nešto veći skup osnovnih operacija. U taj skup ćemo uključiti transformacije $I$, $X$, $Y$, $Z$, $H$, promenu faze $T_{\pi/8}$, $CNOT$ i ostala kontrolisana kola, $SWAP$ i $CCNOT$.

Kvantna merenja u odnosu na bilo koju bazu se mogu svesti na kvantna merenja u odnosu na kanonsku bazu. Zbog toga smatramo da je na kvantnom računaru implementirano jedino merenje u odnosu na kanonsku bazu.

## Kvantna teleportacija

*Kvantna teleportacija* predstavlja postupak slanja kvantnog stanja jednog kubita sa jednog kvantnog računara na drugi kvantni računar preko klasičnog kanala (kanala preko kojeg se šalju klasični bitovi).

Algoritam koji sledi je predstavljen 1993. od strane Čarlsa Beneta (eng. *Charles H. Bennett*), Žila Brasara (fr. *Gilles Brassard*), Kloda Krepoa (fr. *Claude Crépeau*), Ričarda Jože (eng. *Richard Jozsa*), Ašera Peresa (eng. *Asher Peres*) i Vilijama Vutersa (eng. -*William K. Wootters*). Pretpostavke su sledeće:

- postoje dva aktera: Alisa i Boban;
- Alisa poseduje kubit $a_0 \ket{0} + a_1 \ket{1}$;
- Alisa i Boban poseduju svako po jedan kubit tako da je sistem od ta dva kubita u stanju $\ket{\Phi}^{+}$. Skraćeno ćemo reći da Alisa i Boban \textit{dele} sistem $\ket{\Phi}^{+}$, a njegovo kvantno stanje zapisujemo u obliku $\frac{1}{\sqrt{2}} (\ket{0}_A \ket{0}_B) + \ket{1}_A \ket{1}_B)$, gde indeks označava kome pripada pojedinačni kubit u sistemu;
- Alisa i Boban smeju da primenjuju jedino kvantno merenje i unitarne transformacije nad kubitima koji se nalaze na njihovim računarima. Za takve operacije se kaže da su *lokalne*;
- Dozovljena je komunikacija između Alise i Bobana klasičnim kanalom.

Cilj algoritma je da Boban na kraju poseduje kubit čije je kvantno stanje isto kao Alisin kubit $a_0 \ket{0} + a_1 \ket{1}$.

Stanje sistema od sva tri kubita koje Alisa i Boban poseduju se može zapisati na sledeći način:
\begin{align*}
    \frac{1}{\sqrt{2}} (a_0 \ket{0}_A + a_1 \ket{1}_A) \otimes (\ket{0}_A \ket{0}_B + \ket{1}_A \ket{1}_B).
\end{align*}

Algoritam se sastoji od narednih koraka:

1. Alisa primenjuje transformaciju $CNOT$ na svoja dva kubita, odnosno $CNOT \otimes I$ na celom sistemu. Ova operacija je lokalna, jer ne utiče na Bobanov kubit.

    Rezultat ove transformacije je:
    \begin{align*}
        \frac{a_0}{\sqrt{2}} \ket{0}_A \otimes (\ket{0}_A \ket{0}_B + \ket{1}_A \ket{1}_B) + \frac{a_1}{\sqrt{2}} \ket{1}_A \otimes (\ket{1}_A \ket{0}_B + \ket{0}_A \ket{1}_B).
    \end{align*}
    \item Alisa primenjuje transformaciju $H$ na prvom kubitu, odnosno transformaciju $H \otimes I \otimes I$ na celom sistemu.

    Rezultat ove transformacije je:
    \begin{align*}
        &\frac{a_0}{2} (\ket{0}_A + \ket{1}_A) (\ket{0}_A \ket{0}_B + \ket{1}_A \ket{1}_B)
        + \frac{a_1}{2} (\ket{0}_A - \ket{1}_A) (\ket{1}_A \ket{0}_B + \ket{0}_A \ket{1}_B), \\
        = &\frac{1}{2} \ket{00}_A (a_0 \ket{0}_B + a_1 \ket{1}_B) \\
        + &\frac{1}{2} \ket{01}_A (a_0 \ket{1}_B + a_1 \ket{0}_B) & \\
        + &\frac{1}{2} \ket{10}_A (a_0 \ket{0}_B - a_1 \ket{1}_B) & \\
        + &\frac{1}{2} \ket{11}_A (a_0 \ket{1}_B - a_1 \ket{0}_B).
    \end{align*}

2. Alisa vrši merenje u odnosu na kanonsku bazu u $\mathbb{C}^2 \otimes \mathbb{C}^2$ i rezultat $(a, b)$ šalje Bobanu kao dva klasična bita.

    U narednoj tabeli je ilustrovan svaki od mogućih scenarija.

   | $(a, b)$ | Bobanov kubit                   |
   |:--------:|:-------------------------------:|
   | $00$     | $a_0 \ket{0}_B + a_1 \ket{1}_B$ |
   | $01$     | $a_0 \ket{1}_B + a_1 \ket{0}_B$ |
   | $10$     | $a_0 \ket{0}_B - a_1 \ket{1}_B$ |
   | $11$     | $a_0 \ket{1}_B - a_1 \ket{0}_B$ |

4. Ako je $b = 1$, Boban najpre primenjuje $X$ nad svojim bitom. Ako je $a = 1$, nakon toga će primeniti $Z$ transformaciju.

    U tabeli iznad možemo videti da, kada je $b = 1$, skalari nisu uz odgovarajuća stanja. $X$ kolo će zameniti mesta stanjima $\ket{0}$ i $\ket{1}$. Nakon toga, stanje Bobanovog kubita će biti kao u tabeli ispod:
|$(a, b)$ | Bobanov kubit                   |
|:-------:|:-------------------------------:|
|$00$     | $a_0 \ket{0}_B + a_1 \ket{1}_B$ | 
|$01$     | $a_0 \ket{0}_B + a_1 \ket{1}_B$ |
|$10$     | $a_0 \ket{0}_B - a_1 \ket{1}_B$ |
|$11$     | $a_0 \ket{0}_B - a_1 \ket{1}_B$ |

Sada, za $a = 1$, vidimo da je znak uz $\ket{1}$ negativan, što se popravlja primenom $Z$ kola. Na taj način će Boban u sva četiri slučaja uspeti da rekreira Alisin kubit, čime je teleportacija izvršena.

### Q# i kvantna teleportacija

Sada ćemo videti kako se ovo može implementirati u Q#. Kao i kod protokola **BB84**, ova implementacija upotrebljiva u praksi, niti možemo implementirati algoritam tako da on bude upotrebljiv u praksi, jer odgovarajuća tehnologija nam još nije dostupna. Ono što predstavlja problem je nameštanje situacije u kojoj Alisa i Boban dele stanje $\ket{Phi^+}$. Trenutno ne možemo na svojim računarima to da izvedemo, ali vredi napomenuti da su naučnici uspeli to da reše eksperimentalno tako što je Alisa napravila to kvantno stanje, nakon čega je prosledila Bobanu jedan kubit. Više o tome možete naći u sledećem [članku](https://www.innovationnewsnetwork.com/quantum-teleportation-breakthrough-advances-quantum-computing/55351/).

In [1]:
import qsharp

In [2]:
%%qsharp

open Microsoft.Quantum.Diagnostics;

// prva pretpostavka algoritma: Alisa i Boban dele sistem u Belovom stanju
// za potrebe testiranja, definišaćemo funkciju koja će od dva kubita napraviti taj sistem

operation makeBellState(aliceQubit : Qubit, bobQubit : Qubit) : Unit {
    // prebacujemo oba kubita u stanje |0>
    Reset(aliceQubit);
    Reset(bobQubit);
    H(aliceQubit);
    CNOT(aliceQubit, bobQubit);
}

// sada možemo definisati funkciju teleportacije
operation TeleportMessage(message : Qubit, aliceQubit : Qubit, bobQubit : Qubit) : Unit {

    // korak 1: Alisa primenjuje CNOT na svoja dva kubita
    CNOT(message, aliceQubit);

    // korak 2: Alisa primenjuje H na svom kubitu
    H(message);

    // korak 3: Alisa vrši kvantno merenje na svoja dva kubita
    let m1 = M(message);
    let m2 = M(aliceQubit);

    // korak 4: Boban koristi rezultate Alisinog merenja da u svoj kubit upiše Alisinu poruku
    if m2 == One {
        X(bobQubit);
    }

    if m1 == One {
        Z(bobQubit);
    }
    
}

use A = Qubit();
use B = Qubit();

// uzećemo da je početno stanje poruke |+>
use mess = Qubit();
H(mess);

// obezbeđujemo da Alisa i Boban dele sistem u Belovom stanju
makeBellState(A, B);


Message("Checkpoint 1");
DumpMachine();

TeleportMessage(mess, A, B);

Message("Checkpoint 2");
DumpMachine();

Checkpoint 1

<table class="qs-stateTable">
  <style>
    .qs-stateTable thead tr {
      background-color: var(
        --vscode-list-hoverBackground,
        var(--jp-layout-color1, inherit)
      );
    }
    .qs-stateTable th {
      text-align: left;
      border: none;
    }
    .qs-stateTable tbody {
      pointer-events: none;
    }
    .qs-stateTable tbody td {
      text-align: left;
      border: none;
    }
    .qs-stateTable tbody td span {
      display: inline-block;
    }
    .qs-stateTable tbody tr:nth-child(even) {
      background-color: var(
        --vscode-list-hoverBackground,
        var(--jp-layout-color1, inherit)
      );
    }
  </style>
  <thead>
    <tr>
      <th>Basis State<br />(|𝜓₁…𝜓ₙ⟩)</th>
      <th>Amplitude</th>
      <th>Measurement Probability</th>
      <th colspan="2">Phase</th>
    </tr>
  </thead>
  <tbody>
    <tr>
  <td>
    <span>|000⟩</span>
  </td>
  <td>
    <span>0.5000+0.0000𝑖</span>
  </td>
  <td>
    <progress max="100" value="25.00000000000001"></progress>
    <span>25.0000%</span>
  </td>
  <td style="transform: rotate(0.0000rad)">↑</td>
  <td>
    <span>0.0000</span>
  </td>
</tr>
<tr>
  <td>
    <span>|001⟩</span>
  </td>
  <td>
    <span>0.5000+0.0000𝑖</span>
  </td>
  <td>
    <progress max="100" value="25.00000000000001"></progress>
    <span>25.0000%</span>
  </td>
  <td style="transform: rotate(0.0000rad)">↑</td>
  <td>
    <span>0.0000</span>
  </td>
</tr>
<tr>
  <td>
    <span>|110⟩</span>
  </td>
  <td>
    <span>0.5000+0.0000𝑖</span>
  </td>
  <td>
    <progress max="100" value="25.00000000000001"></progress>
    <span>25.0000%</span>
  </td>
  <td style="transform: rotate(0.0000rad)">↑</td>
  <td>
    <span>0.0000</span>
  </td>
</tr>
<tr>
  <td>
    <span>|111⟩</span>
  </td>
  <td>
    <span>0.5000+0.0000𝑖</span>
  </td>
  <td>
    <progress max="100" value="25.00000000000001"></progress>
    <span>25.0000%</span>
  </td>
  <td style="transform: rotate(0.0000rad)">↑</td>
  <td>
    <span>0.0000</span>
  </td>
</tr>

  </tbody>
</table>


$|\psi\rangle = \frac{1}{2}|000\rangle+\frac{1}{2}|001\rangle+\frac{1}{2}|110\rangle+\frac{1}{2}|111\rangle$

Checkpoint 2

<table class="qs-stateTable">
  <style>
    .qs-stateTable thead tr {
      background-color: var(
        --vscode-list-hoverBackground,
        var(--jp-layout-color1, inherit)
      );
    }
    .qs-stateTable th {
      text-align: left;
      border: none;
    }
    .qs-stateTable tbody {
      pointer-events: none;
    }
    .qs-stateTable tbody td {
      text-align: left;
      border: none;
    }
    .qs-stateTable tbody td span {
      display: inline-block;
    }
    .qs-stateTable tbody tr:nth-child(even) {
      background-color: var(
        --vscode-list-hoverBackground,
        var(--jp-layout-color1, inherit)
      );
    }
  </style>
  <thead>
    <tr>
      <th>Basis State<br />(|𝜓₁…𝜓ₙ⟩)</th>
      <th>Amplitude</th>
      <th>Measurement Probability</th>
      <th colspan="2">Phase</th>
    </tr>
  </thead>
  <tbody>
    <tr>
  <td>
    <span>|001⟩</span>
  </td>
  <td>
    <span>0.7071+0.0000𝑖</span>
  </td>
  <td>
    <progress max="100" value="49.999999999999986"></progress>
    <span>50.0000%</span>
  </td>
  <td style="transform: rotate(0.0000rad)">↑</td>
  <td>
    <span>0.0000</span>
  </td>
</tr>
<tr>
  <td>
    <span>|011⟩</span>
  </td>
  <td>
    <span>0.7071+0.0000𝑖</span>
  </td>
  <td>
    <progress max="100" value="49.999999999999986"></progress>
    <span>50.0000%</span>
  </td>
  <td style="transform: rotate(0.0000rad)">↑</td>
  <td>
    <span>0.0000</span>
  </td>
</tr>

  </tbody>
</table>


$|\psi\rangle = \frac{\sqrt{2}}{2}|001\rangle+\frac{\sqrt{2}}{2}|011\rangle$

Kao što vidimo na *Checkpoint 1*, kvantni sistem $A \otimes B \otimes mess$ je u stanju $\ket{\Phi^+} \otimes \ket{+}$. Nakon izvršavanja algoritma, kao što vidimo u *Checkpoint 2*, sistem je u stanju $\ket{1} \otimes \ket{+} \otimes \ket{1}$. Dakle, Bobanov kubit $B$ je sada u istom kvantnom stanju kao poruka $mess$ pre izvršavanja algoritma.

## Dojčov problem

Sada ćemo prikazati algoritam koji je eksponencijalno puta brži na kvantnom računaru nego na klasičnom. Algoritam, koji je dobio naziv po engleskom naučniku Dejvidu Dojču (eng. *David Deutsch*), rešava sledeći problem: za datu funkciju $f: \mathbb{Z}_2 \xrightarrow{} \mathbb{Z}_2$, odrediti da li je $f$ bijekcija, odnosno da li je $f(0) \neq f(1)$ ili je $f(0) = f(1)$.

Kako bismo pristupili problemu, najpre ćemo posmatrati sledeću unitarnu transformaciju:
\begin{align*}
    \sqcup_f \ket{x}\ket{y} = \ket{x}\ket{y +_2 f(x)},
\end{align*} gde su $x, y$ iz skupa $\{0, 1\}$. Kao što smo videli u prethdnoj lekciji, ova unitarna transformacija predstavlja kvantni algoritam za izračunavanje funkcije $f(x)$.  Može se pretpostaviti da je $\sqcup_f$ već dato kao *crna kutija* i da je potrebno, sa što manje poziva te funkcije (primene transformacije $\sqcup_f$) otkriti da li je $f$ bijekcija ili nije.

Pre nego što predstavimo algoritam, daćemo primer ovakvog preslikavanja.

#### Primer transformacije $\sqcup_f$:

Neka je data funkcija $f : \mathbb{Z}_2 \xrightarrow[]{} \mathbb{Z}_2$ za koju važi $f(x) = 1$, za $x \in \mathbb{Z}_2$.

Dakle, $\sqcup_f \ket{x} \ket{y} = \ket{x} \ket{y +_2 1}$. Kao što je poznato u linearnoj algebri, da bismo dobili matricu koja predstavlja linearno preslikavanje $\sqcup_f$, dovoljno je da odredimo rezultat tog preslikavanja na baznim vektorima. Odatle dobijamo sledeće:

\begin{align*}
        \sqcup_f \ket{00} = \ket{01}, \\
        \sqcup_f \ket{01} = \ket{00}, \\
        \sqcup_f \ket{10} = \ket{11}, \\
        \sqcup_f \ket{11} = \ket{10}.
\end{align*}

Budući da je $\ket{0} = (1, 0)^T$, a $\ket{1} = (0, 1)^T$, dobijamo:
\begin{align*}
        \ket{00} = (1, 0, 0, 0)^T, \\
        \ket{01} = (0, 1, 0, 0)^T, \\
        \ket{10} = (0, 0, 1, 0)^T, \\
        \ket{11} = (0, 0, 0, 1)^T.
\end{align*}

Dakle, $\sqcup_f = \begin{pmatrix}
        0 & 1 & 0 & 0 \\
        1 & 0 & 0 & 0 \\
        0 & 0 & 0 & 1 \\
        0 & 0 & 1 & 0
    \end{pmatrix}$.

Primetimo da je ovo preslikavanje ekvivalentno sa $I \otimes X$.

Algoritam se zasniva na jednom *triku*. Ako postavimo $x \in \{0, 1\}$, $y = \ket{-} = \frac{1}{\sqrt{2}}(\ket{0} - \ket{1})$, dobijamo:
\begin{align*}
    \sqcup_f \ket{x} \ket{-} = & \frac{1}{\sqrt{2}} \sqcup_f \ket{x} \ket{0} - \frac{1}{\sqrt{2}} \sqcup_f \ket{x} \ket{1} \\
    = & \frac{1}{\sqrt{2}} \ket{x} \ket{f(x)} - \frac{1}{\sqrt{2}} \ket{x} \ket{f(x) +_2 1} \\
    = & (-1)^{f(x)} \ket{x}\ket{-}.
\end{align*}

Ako postavimo sada $x = \ket{+}$, dobijamo:
\begin{align*}
    \sqcup_f \ket{+} \ket{-} = & \frac{1}{\sqrt{2}} \sqcup_f \ket{0}\ket{-} + \frac{1}{\sqrt{2}} \sqcup_f \ket{1}\ket{-} \\
    = & \frac{1}{\sqrt{2}}\left( (-1)^{f(0)} \ket{0}\ket{-} + (-1)^{f(1)} \ket{1}\ket{-} \right) \\
    = & \frac{(-1)^{f(0)}}{\sqrt{2}} \left(\ket{0}\ket{-} + (-1)^{f(0) +_2 f(1)} \ket{1} \ket{-} \right).
\end{align*}

Ako je $f$ bijekcija, biće $f(0) +_2 f(1) = 1$, a u suprotnom će biti $f(0) +_2 f(1) = 0$. Kada je $f(0) +_2 f(1) = 1$, $\sqcup_f \ket{+} \ket{-} = (-1)^{f(0)} \ket{-} \ket{-}$; za $f(0) +_2 f(1) = 0$, $\sqcup_f \ket{+} \ket{-} = (-1)^{f(0)} \ket{+} \ket{-}$. Odavde sledi da možemo, nakon primene $\sqcup_f$ na ulazu $\ket{+}\ket{-}$, da razlikujemo da li je $f$ bijekcija ili nije, tako što primenimo kvantno merenje u Adamarovoj bazi na prvom kubitu.

#### Primer:

Neka je $\sqcup_f$ kao u prethodnom primeru. Odatle sledi:
    \begin{align*}
        \sqcup_f \ket{+} \ket{-} &= \begin{pmatrix}
        0 & 1 & 0 & 0 \\
        1 & 0 & 0 & 0 \\
        0 & 0 & 0 & 1 \\
        0 & 0 & 1 & 0
    \end{pmatrix} \left( \begin{pmatrix}
        \frac{1}{\sqrt{2}} \\  \frac{1}{\sqrt{2}}
    \end{pmatrix} \otimes \begin{pmatrix}
        \frac{1}{\sqrt{2}} \\ -\frac{1}{\sqrt{2}}
    \end{pmatrix} \right) \\
    &= \begin{pmatrix}
        0 & 1 & 0 & 0 \\
        1 & 0 & 0 & 0 \\
        0 & 0 & 0 & 1 \\
        0 & 0 & 1 & 0
    \end{pmatrix} \begin{pmatrix}
        \frac{1}{2} \\ -\frac{1}{2} \\ \frac{1}{2} \\ -\frac{1}{2}
    \end{pmatrix} = \begin{pmatrix}
        -\frac{1}{2} \\ \frac{1}{2} \\ -\frac{1}{2} \\ \frac{1}{2}
    \end{pmatrix} = -\ket{+}\ket{-}.
    \end{align*}

Obično se algoritam realizuje ovako:
\begin{enumerate}
1. Početno stanje je $\ket{00}$.
2. Primeni Adamarovu transformaciju na prvom kubitu, a na drugom primeni Paulijevu $X$ transformaciju. Sada je stanje $\ket{+}\ket{1}$.
3. Primeni Adamarovu transformaciju na drugom kubitu. Sada je stanje $\ket{+}\ket{-}$.
4. Primeni transformaciju $\sqcup_f$ na oba kubita. Sada je stanje $\ket{+}\ket{-}$ ako $f$ nije bijekcija, a inače je $\ket{-}\ket{-}$.
5. Primeni Adamarovu transformaciju na prvom kubitu. Ako $f$ nije bijekcija, $\sqcup_f \ket{+}\ket{-} = (-1)^{f(0)} \ket{+}\ket{-}$. Primenom Adamarove transformacije na prvom kubitu, stanje ovog sistema postaje $(-1)^{f(0)} \ket{0}\ket{-}$. Slično, ako je $f$ bijekcija, stanje sistema postaje $(-1)^{f(0)} \ket{1}\ket{-}$.
6. Primeni merenje u kanonskoj bazi. Ako je rezultat $0$, $f$ nije bijekcija, a inače jeste.

Na klasičnom računaru su potrebna dva poziva funkcije $f$ da bi se ovo odredilo, dok je na kvantnom računaru dovoljan jedan poziv unitarne transformacije $\sqcup_f$. Dakle, kvantni algoritam radi sa duplo manje poziva funkcije $f$, odnosno $\sqcup_f$.

### Q# i Dojčov problem

In [3]:
%%qsharp

operation HadamardMeasurement(q : Qubit) : Result {
    H(q);
    return M(q);
}

operation Deutsch(U : (Qubit, Qubit) => Unit is Adj) : Bool {
    use q1 = Qubit();
    use q2 = Qubit();

    // q1 = |0>  --> |+>
    H(q1);
    // q2 = |0> --> |1> --> |->
    X(q2);
    H(q2);

    U(q1, q2);

    let m = HadamardMeasurement(q1);

    // pre završetka funkcije, moramo resetovati kubite
    Reset(q1);
    Reset(q2);

    return m == One;

}

operation U1(q1 : Qubit, q2 : Qubit) : Unit is Adj {
    X(q2);  
}


Message($"U1 je bijekcija: {Deutsch(U1)}");

U1 je bijekcija: false

Pokušajmo sada sa funkcijom koja jeste bijekcija, recimo $f(0) = 1$, $f(1) = 0$. Nju opisuje transformacija:
\begin{align*}
\sqcup_f \ket{00} = \ket{01}, \\
\sqcup_f \ket{01} = \ket{00}, \\
\sqcup_f \ket{10} = \ket{10}, \\
\sqcup_f \ket{11} = \ket{11}.
\end{align*}

Ova transformacija je ekvivalentna sa $(X \otimes I)(CNOT(X \otimes I))$.

In [4]:
%%qsharp

operation U2(q1 : Qubit, q2 : Qubit) : Unit is Adj {
    X(q1);
    CNOT(q1, q2);
    X(q1);
}

Message($"U2 je bijekcija: {Deutsch(U2)}");

U2 je bijekcija: true

## Algoritam Dojča i Jože

Ovaj algoritam, koji je dobio naziv po Dejvidu Dojču i profesoru Ričardu Joži, rešava opštiji slučaj Dojčovog problema. Funkcija $f: \{0, 1\}^n \xrightarrow{} \{0, 1\}$ sada može uzeti vrednosti od $0$ do $2^{n - 1}$, za neko $n$ i ona je ili konstantna, ili je *balansirana*, odnosno $2^{n - 1}$ vrednosti je $0$, a $2^{n - 1}$ vrednosti je $1$.

Originalni algoritam je zahtevao dva poziva funkcije $\sqcup_f$, ali je kasnije pokazano kako se broj poziva može smanjiti na jedan.

Osnovna ideja se zasniva na slučaju $n = 1$. Ako posmatramo niz kubita $\ket{x_i}$, gde je $x_i \in \{0, 1\}^{n}$ kao stanje $\ket{i}$, gde je $x_i$ binarni zapis broja $i$, ispostavlja se da važi:
\begin{align*}
    \sqcup_f \ket{i}\ket{-} = (-1)^{f(i)} \ket{i} \ket{-}.
\end{align*}

Na osnovu ovoga se pokazuje da se prethodni algoritam može uopštiti na sledeći način:
1. Početno stanje je $\ket{0}^{\otimes (n + 1)}$.
2. Primeni Adamarovu transformaciju na niz od prvih $n$ kubita, a na poslednjem primeni Paulijevu $X$ transformaciju. Sada je stanje $\ket{+}^{\otimes n} \ket{1}$
3. Primeni Adamarovu transformaciju na poslednjem kubitu od $\ket{+}^{\otimes n} \ket{1}$, čime dobijaš stanje $\ket{+}^{\otimes n} \ket{-}$
4. Primeni transformaciju $\sqcup_f$ na svim kubitima.
5. Primeni Adamarovu transformaciju na nizu od prvih $n$ kubita. Ako $f$ nije balansirana, stanje je $\ket{0}^{\otimes n}\ket{-}$.
6. Primeni merenje u kanonskoj bazi. Ako je rezultat $0$, $f$ nije balansirana, a inače jeste.

Ovde možemo primetiti da je potreban samo jedan poziv funkcije $\sqcup_f$ i $O(n)$ ostalih operacija, dok je na klasičnom računaru potrebno bar $2^{n - 1} + 1$ poziva funkcije $f$ da bi se odredilo da li je $f$ balansirana. Jedan od razloga za ovoliko ubrzanje dolazi od toga što na kvantnim računarima jedan kubit predstavlja veći broj podataka nego što to čini jedan bit u klasičnom računaru. Zbog toga smo, analizom rada funkcije $\sqcup_f$ na nizu kubita $\ket{+}^{\otimes(n - 1)}\ket{-}$ mogli da odredimo njeno ponašanje na kubitima $\ket{x_i}$.

### Q# i algoritam Dojča i Jože

Implementacija sledi iz opisa algoritma. Testiraćemo algoritam na istom ulazu koji smo koristili kad smo testirali Dojčov.

In [5]:
%%qsharp

operation DeutschJozsa(n : Int, U : Qubit[] => Unit) : Bool {
    // korak 1: početno stanje je |0>^{n + 1}
    use qubits = Qubit[n + 1];

    // korak 2: Adamar na svima, osim na poslednjem; na poslednjem X
    for i in 0..n-1 {
        H(qubits[i]);
    }
    X(qubits[n]);

    // korak 3: Adamar na poslednjem
    H(qubits[n]);

    // korak 4: Uf
    U(qubits);

    // korak 5: Adamar na prvih n
    for i in 0..n-1 {
        H(qubits[i]);
    }

    // korak 6: merenje u kanonskoj bazi
    mutable r = 0;
    for i in 0..n-1 {
        r *= 2;
        if M(qubits[i]) == One {
            r += 1;
        }
    }

    // moramo resetovati sve kubite
    for i in 0..n {
        Reset(qubits[i]);
    }

    // ako je r = 0, onda nije balansirana; inače jeste
    return r != 0;
}

operation U3(qubits : Qubit[]) : Unit {
    U1(qubits[0], qubits[1]);
}

operation U4(qubits : Qubit[]) : Unit {
    U2(qubits[0], qubits[1]);
}


Message($"U3 je balansirana: {DeutschJozsa(1, U3)}");
Message($"U4 je balansirana: {DeutschJozsa(1, U4)}");

U3 je balansirana: false

U4 je balansirana: true

Možda ne bi bilo loše da testiramo algoritam na malo većoj funkciji. Probaćemo na funkciji $f : \{0, 1\}^2 \xrightarrow{} \{0, 1\}$. Neka je $f(0) = 0$, $f(1) = 1$, $f(2) = 0$, $f(3) = 1$. Odatle je $\sqcup_f$ definisana narednim jednačinama:
\begin{align*}
\sqcup_f \ket{000} = \ket{000}, \\
\sqcup_f \ket{001} = \ket{001}, \\
\sqcup_f \ket{010} = \ket{011}, \\
\sqcup_f \ket{011} = \ket{010}, \\
\sqcup_f \ket{100} = \ket{100}, \\
\sqcup_f \ket{101} = \ket{101}, \\
\sqcup_f \ket{110} = \ket{111}, \\
\sqcup_f \ket{111} = \ket{110}, \\
\end{align*}

Sada ćemo predstaviti $\sqcup_f$ preko elementarnih transformacija. Primetimo da se stanje trećeg bita menja ako je stanje prva dva $\ket{01}$ ili $\ket{11}$. Ako primenimo $CCNOT$, rešili smo drugi slučaj. Ako primenimo $X \otimes I \otimes I$, pa $CCNOT$, pa $X \otimes I \otimes I$, rešili smo i prvi slučaj. Dakle, tvrdimo da naredni kod opisuje funkciju $\sqcup_f$.

In [6]:
%%qsharp

operation U5(qubits : Qubit[]) : Unit {
    let x1 = qubits[0];
    let x2 = qubits[1];
    let y = qubits[2];

    CCNOT(x1, x2, y);
    X(x1);
    CCNOT(x1, x2, y);
    X(x1);
}

Lako se proverava da je ova funkcija ispravna. To se može uraditi tako što se preko DumpMachine() testira rezultat svakog baznog vektora. Potrebno je posmatrati poslednja tri kubita u sistemu, jer prva tri su vezana za testiranje kvantne teleportacije.

In [7]:
%%qsharp

Message($"U5 je balansirana: {DeutschJozsa(2, U5)}");

U5 je balansirana: true